wifi: shell: Support multiple security modes for each network

rado17 opened the pull request:
Support configuration of multiple security modes for each network.

rado17 commented:
@sachinthegreen @VivekUppunda @kapbh @ajayparida Please check

ajayparida commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -490,10 +492,32 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 			break;
 		case 'k':
-			params->security = atoi(optarg);
+			params->security = BIT(WIFI_SECURITY_TYPE_PSK);
```
If the user gives "none", still PSK bit will be set(as it's being set default). Please check.

ajayparida commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1873,7 +1897,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
-		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs)\n"
+		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs). User can provide multiple Security types by giving comma-separated string]\n"
+		  "eg., \"1,2,3\" for \"WPA2-PSK,WPA2-PSK-256,SAE\"\n"
 		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK\n"
```
Please remove a space after "7:  " for others not present.

VivekUppunda commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -490,10 +492,32 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
+			}
+
+			memset(security, 0, strlen(optarg));
+			strcpy(security, optarg);
```
do we use strcpy or strncpy? Can you kindly please check this

rado17 replied:
Used strcpy here since size of `security` buffer is equal to length of optarg string. According to another suggestion that I got from Robert Lubos, will make changes to avoid this completely.

VivekUppunda commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -490,10 +492,32 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
+
+			/* Security type (optional) */
+
+			security = (char *)k_malloc(strlen(optarg));
```
I would agree for the above. After looking at a few heap allocation failure issues in the past couple of weeks, I would wish to have lesser heap allocation code if it is possible.

krish2718 reviewed (changes requested):
As [suggested](https://github.com/zephyrproject-rtos/zephyr/pull/71044/files#r1549652203) iadding a `WPA_AUTO_PERSONAL` would be enough and we can let the drivers/NMs implement this feature.

rado17 committed:
wifi: shell: Support WPA auto personal security mode

Applications need to run a scan to identify the security
mode before attempting the connection which adds to the time
taken for connection to be established.
To avoid the initial scan, support auto security mode which will
enable STA to choose between WPA, WPA2 and WPA3, based on the
network configuration.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>

krish2718 commented on include/zephyr/net/wifi.h:
```diff
@@ -49,6 +49,8 @@ enum wifi_security_type {
 	/** WPA-PSK security. */
 	WIFI_SECURITY_TYPE_WPA_PSK,
+	/** WPA/WPA2/WPA3 PSK security. */
+	WIFI_SECURITY_TYPE_WPA_AUTO,
```
`WPA_AUTO_PERSONAL` might limit to personal security.

krish2718 commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -275,7 +275,8 @@ static int wifi_connect(uint32_t mgmt_request, struct net_if *iface,
 		  params->security == WIFI_SECURITY_TYPE_WPA_PSK ||
-		  params->security == WIFI_SECURITY_TYPE_PSK_SHA256) &&
+		  params->security == WIFI_SECURITY_TYPE_PSK_SHA256 ||
+		  params->security == WIFI_SECURITY_TYPE_WPA_AUTO) &&
```
```suggestion
		  params->security == WIFI_SECURITY_TYPE_WPA_AUTO_PERSONAL) &&
```

rado17 replied:
Done.

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1870,7 +1870,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-p, --psk]: Passphrase (valid only for secure SSIDs)\n"
 		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs)\n"
-		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK\n"
+		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK, 8: WPA-Auto\n"
```
`WPA-Auto-Personal`

rado17 replied:
Done

rado17 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1870,7 +1870,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		  "[-p, --psk]: Passphrase (valid only for secure SSIDs)\n"
 		  "[-k, --key-mgmt]: Key Management type (valid only for secure SSIDs)\n"
-		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK\n"
+		  "0:None, 1:WPA2-PSK, 2:WPA2-PSK-256, 3:SAE, 4:WAPI, 5:EAP, 6:WEP, 7: WPA-PSK, 8: WPA-Auto-Personal\n"
```
Done.

krish2718 replied:
IIRC, String only lines are exempted from line checks.

jukkar replied:
Ok, I did not know that.
